src/libraries/Microsoft.Extensions.Http/src/Logging/HttpClientLoggerHandler.cs

```diff
@@ -0,0 +1,90 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Net.Http;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Internal;
+
+namespace Microsoft.Extensions.Http.Logging
+{
+    internal sealed class HttpClientLoggerHandler : DelegatingHandler
+    {
+        private readonly IHttpClientLogger _httpClientLogger;
+        private readonly IHttpClientAsyncLogger? _httpClientAsyncLogger;
+
+        public HttpClientLoggerHandler(IHttpClientLogger httpClientLogger)
+        {
+            ThrowHelper.ThrowIfNull(httpClientLogger);
+
+            _httpClientLogger = httpClientLogger;
+            _httpClientAsyncLogger = httpClientLogger as IHttpClientAsyncLogger;
+        }
+
+        protected override async Task<HttpResponseMessage> SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
+        {
+            ThrowHelper.ThrowIfNull(request);
+
+            var stopwatch = ValueStopwatch.StartNew();
+            HttpResponseMessage? response = null;
+
+            object? state = _httpClientAsyncLogger is not null
+                ? await _httpClientAsyncLogger.LogRequestStartAsync(request, cancellationToken).ConfigureAwait(false)
+                : _httpClientLogger.LogRequestStart(request);
+
+            try
+            {
+                response = await base.SendAsync(request, cancellationToken).ConfigureAwait(false);
+
+                if (_httpClientAsyncLogger is not null)
+                {
+                    await _httpClientAsyncLogger.LogRequestStopAsync(state, request, response, stopwatch.GetElapsedTime(), cancellationToken).ConfigureAwait(false);
+                }
+                else
+                {
+                    _httpClientLogger.LogRequestStop(state, request, response, stopwatch.GetElapsedTime());
+                }
+                return response;
+            }
+            catch (Exception exception)
+            {
+                if (_httpClientAsyncLogger is not null)
+                {
+                    await _httpClientAsyncLogger.LogRequestFailedAsync(state, request, response, exception, stopwatch.GetElapsedTime(), cancellationToken).ConfigureAwait(false);
```

Comment:
Not sure if we care about this, but this will also try to log if `LogRequestStop(Async)` throws. How about saving `exception` into another `Exception?` variable, and do the logging in the `finally` block?

Reply:
Given that exception was unhandled by the implementation of the logging, it _will_ fail the request... So I don't really see a problem with it being included.

The thing I might consider (though, in a separate PR) is to potentially swallow all the exceptions from the logging but log it to "Private.InternalDiagnostics...." event source. But we don't have the event source yet. It was something discussed in relation to https://github.com/dotnet/runtime/issues/77312#issuecomment-1625715837 and https://github.com/dotnet/runtime/issues/77312#issuecomment-1631322543 and I will create an issue for it in the upcoming days. But it might not make it into 8.0...

In the meantime we should mention in the docs that loggers _should not_ throw any exceptions.

Reply:
Added to the remarks in triple-slash comments